src/metabase/api/database.clj

```diff
@@ -352,6 +353,63 @@
                                      secret/expand-db-details-inferred-secret-values
                                      (assoc :can-manage true)))))
 
+(def ^:private database-usage-models
+  "List of models that are used to report usage on a database."
+  [:question :dataset :metric :segment])
+
+(defmulti ^:private database-usage-query
+  "Query that will returns the number of `model` that use the database with id `database-id`.
+  The query must returns a scalar, and the method could return `nil` in case no query is available."
+  {:arglists '([model database-id table-ids])}
+  (fn [model _database-id _table-ids] (keyword model)))
+
+(defmethod database-usage-query :question
+  [_ db-id _table-ids]
+  {:select [[:%count.* :question]]
+   :from   [:report_card]
+   :where  [:and
+            [:= :database_id db-id]
+            [:= :dataset false]]})
+
+(defmethod database-usage-query :dataset
+  [_ db-id _table-ids]
+  {:select [[:%count.* :dataset]]
+   :from   [:report_card]
+   :where  [:and
+            [:= :database_id db-id]
+            [:= :dataset true]]})
+
+(defmethod database-usage-query :metric
+  [_ _db-id table-ids]
+  (when (seq table-ids)
+    {:select [[:%count.* :metric]]
+     :from   [:metric]
+     :where  [:in :table_id table-ids]}))
+
+(defmethod database-usage-query :segment
+  [_ _db-id table-ids]
+  (when (seq table-ids)
+    {:select [[:%count.* :segment]]
+     :from   [:segment]
+     :where  [:in :table_id table-ids]}))
+
+(api/defendpoint GET "/:id/usage_info"
+  "Get usage info for a database.
+  Returns a map with keys are models and values are the number of entities that use this database."
+  [id]
+  {id ms/IntGreaterThanZero}
+  (api/check-superuser)
+  (api/check-404 (db/exists? Database :id id))
+  (let [table-ids (db/select-ids Table :db_id id)]
+    (merge (into {} (for [model database-usage-models]
+                      [model 0]))
```

Comment:
I'm not sure I can explain my reasoning, but it seems a little cleaner to me to not do the `(merge (into ... [model 0]))` stuff and instead change each `:where` to something like

```
(defn- table-id-clause
  [table-ids]
  (if (seq table-ids)
    [:in table_id table-ids]
    [:= 0 1]))
    
(defmethod database-usage-query :whatever
  ...
  :where (table-id-clause table-ids))
  ```
  
  But if you have strong feelings against that I certainly won't hold you to it; do what you want.

Reply:
hmm, but I guess that's a rare case. so it's fine to waste some DB cycle. 
will switch back.

Reply:
It's still just a single query (since all the subqueries get merged together), hard to be too upset about it

Reply:
https://github.com/metabase/metabase/pull/27646/commits/14491e56529301c3482e1a41ea404012d2b373fb